### PR TITLE
feat: make versioned fluent resources linker friendly

### PR DESCRIPTION
**Pull request by jeromelaban**
## PR Type

What kind of change does this PR introduce?
- Feature

## What is the new behavior?

Ensures that Fluent styles versions can be linked out based on usage in App.xaml.

- The change moves styles per version into separate assemblies so that the linker can exclude the content based on usage. 
- The XAML generator detects uses of the `Microsoft.UI.Xaml.Controls.XamlControlsResources` class and rewrites it to either `Microsoft.UI.Xaml.Controls.XamlControlsResourcesV1` or `Microsoft.UI.Xaml.Controls.XamlControlsResourcesV2` depending on the use of the `ControlsResourcesVersion` property, or the default current version (2 as of this PR).

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by MartinZikmund** [comment]
@jeromelaban Would it be possible to move `XamlControlsResources to Uno.UI? Some of the controls in WinUI now check for the ControlsResourcesVersion or other properties on this class, so we will eventually need to have it accessible there (in some way)

File: src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Skia.csproj [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.Skia.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Skia.csproj:29]: Consider marking it as PrivateAssets = "All". I don't think this reference is needed in runtime but only in buildtime.

File: src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Wasm.csproj [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.Wasm.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Wasm.csproj:30]: Same of above

File: src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.csproj [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	
+	<ItemGroup Condition="$(IsMonoAndroid) or $(IsXamarinIOS) or $(IsXamarinMac)">
+		<Reference Include="System.Numerics" />
+		<Reference Include="System.Numerics.Vectors" />
+		<Reference Include="System.Json" />
+		<Reference Include="System.Memory" />
+	</ItemGroup>
+
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.csproj:49]: Same of above

File: src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.net6.csproj [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.net6.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.net6.csproj:32]: Same of above

File: src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.Skia.csproj [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.Skia.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.Skia.csproj:29]: Same of above

File: src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.Wasm.csproj [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.Wasm.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.Wasm.csproj:30]: Same of above

File: src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.csproj [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	
+	<ItemGroup Condition="$(IsMonoAndroid) or $(IsXamarinIOS) or $(IsXamarinMac)">
+		<Reference Include="System.Numerics" />
+		<Reference Include="System.Numerics.Vectors" />
+		<Reference Include="System.Json" />
+		<Reference Include="System.Memory" />
+	</ItemGroup>
+
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.csproj:49]: Same of above

File: src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.net6.csproj [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+		<UseCommonOverridePackage>true</UseCommonOverridePackage>
+		<CommonOverridePackageId>Uno.UI</CommonOverridePackageId>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Uno.UI\Uno.UI.net6.csproj" />
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="Uno.SourceGenerationTasks" />
```
Comment (workgroupengineering) [inline src/Uno.UI.FluentTheme.v2/Uno.UI.FluentTheme.v2.net6.csproj:32]: Same of above

**Comment by jeromelaban** [comment]
> @jeromelaban Would it be possible to move `XamlControlsResources to Uno.UI? Some of the controls in WinUI now check for the ControlsResourcesVersion or other properties on this class, so we will eventually need to have it accessible there (in some way)

How do those classes check this value? Is it for the max version? I don't think there's a way to access an instance otherwise (except by browsing the dictionaries tree).

File: src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Skia.csproj (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.UI.FluentTheme.v1/Uno.UI.FluentTheme.v1.Skia.csproj:29]: Uno packages these references without the nuget "pack" target which uses the PrivateAssets attribute. The modification is not needed.

**Comment by MartinZikmund** [comment]
> > @jeromelaban Would it be possible to move `XamlControlsResources to Uno.UI? Some of the controls in WinUI now check for the ControlsResourcesVersion or other properties on this class, so we will eventually need to have it accessible there (in some way)
> 
> How do those classes check this value? Is it for the max version? I don't think there's a way to access an instance otherwise (except by browsing the dictionaries tree).

Example here - https://github.com/microsoft/microsoft-ui-xaml/blob/94cd32694bcab9dfede50074c632149eb28e5be6/dev/TabView/TabViewItem.cpp#L138 . But I noticed it is just about this one property for now, maybe it will just be easier to just set some static property for this purpose.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlConstants.cs [hunk 9]
```diff
@@ -19,6 +19,8 @@ internal static class XamlConstants
 		public const string BaseXamlNamespace = "Windows.UI.Xaml";
 		public const string UnoXamlNamespace = "Windows.UI.Xaml";
 
+		public const int MaxFluentResourcesVersion = 2;
+
```
Comment (carldebilly) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlConstants.cs:23]: Maybe this should be define somewhere _closer_ to the fluent projects?

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs [hunk 10]
```diff
@@ -72,8 +72,11 @@ internal partial class XamlCodeGeneration
 		private bool _skipUserControlsInVisualTree = false;
 		private readonly GeneratorExecutionContext _generatorContext;
 
-		private bool IsUnoAssembly => _defaultNamespace == "Uno.UI";
-		private bool IsUnoFluentAssembly => _defaultNamespace == "Uno.UI.FluentTheme";
+		private bool IsUnoAssembly
+			=> _defaultNamespace == "Uno.UI";
+
+		private bool IsUnoFluentAssembly
+			=> _defaultNamespace == "Uno.UI.FluentTheme" || _defaultNamespace.StartsWith("Uno.UI.FluentTheme.v");
```
Comment (davidjohnoliver) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:79]: 🤖

**Review by carldebilly — APPROVED** [review]

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 11]
```diff
@@ -1366,6 +1362,49 @@ private void InitializeAndBuildResourceDictionary(IIndentedStringBuilder writer,
...
+							}
+
+							return explicitVersion;
+						}
+					}
+
+					return XamlConstants.MaxFluentResourcesVersion;
+				}
+
+				using (writer.BlockInvariant($"new global::Microsoft.UI.Xaml.Controls.XamlControlsResourcesV{GetResourcesVersion()}()"))
```
Comment (davidjohnoliver) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:1394]: well... fair enough :p

File: src/Uno.UI.FluentTheme.v1/XamlControlsResourcesV1.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+using Uno;
+using Uno.Extensions;
+using Uno.UI.Xaml;
+using Windows.UI.Xaml;
+
+namespace Microsoft.UI.Xaml.Controls
+{
+	public sealed partial class XamlControlsResourcesV1 : ResourceDictionary
```
Comment (davidjohnoliver) [inline src/Uno.UI.FluentTheme.v1/XamlControlsResourcesV1.cs:10]: Is it possible for this to inherit from `XamlControlsResources`? I am thinking about code-behind which eg may look for that type in Application.Resources.

File: src/Uno.UI.FluentTheme.v2/XamlControlsResourcesV2.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+using Uno;
+using Uno.Extensions;
+using Uno.UI.Xaml;
+using Windows.UI.Xaml;
+
+namespace Microsoft.UI.Xaml.Controls
+{
+	public sealed partial class XamlControlsResourcesV2 : ResourceDictionary
```
Comment (davidjohnoliver) [inline src/Uno.UI.FluentTheme.v2/XamlControlsResourcesV2.cs:10]: same here.

**Review by dr1rrb — APPROVED** [review]

**Review by davidjohnoliver — COMMENTED** [review]
Should

**Review by MartinZikmund — APPROVED** [review]

File: src/Uno.UI.FluentTheme.v1/XamlControlsResourcesV1.cs (hunk 12 above)
Comment (jeromelaban) [reply src/Uno.UI.FluentTheme.v1/XamlControlsResourcesV1.cs:10]: Unfortunately, referencing `XamlControlsResources will force an unconditional reference to v1 and v2, and make the linker keep both versions.

**Review by davidjohnoliver — APPROVED** [review]
Since it's not possible to inherit from `XamlControlsResources` for now, should be marked as a breaking change via Conventional Commit

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs (hunk 10 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:79]: Will fix in subsequent PR

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlConstants.cs (hunk 9 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlConstants.cs:23]: We'll change it once v3 is available. The problem here is the availability of this variable from two scopes (inside the generators and in the fluent assemblies), pulling the value is more trouble than it's worth for now.

**Pull request merged**